mllib/src/main/scala/org/apache/spark/ml/feature/RFormula.scala

```diff
@@ -130,9 +173,52 @@ class RFormula(override val uid: String)
       "Label column already exists and is not of type DoubleType.")
   }
 
-  private def hasLabelCol(schema: StructType): Boolean = {
-    schema.map(_.name).contains($(labelCol))
+  private def featureTransformer(schema: StructType): Transformer = {
+    // StringType terms and terms representing interactions need to be encoded before assembly.
+    // TODO(ekl) add support for feature interactions
+    var encoderStages = Seq[Transformer]()
+    var tempColumns = Seq[String]()
+    val encodedTerms = parsedFormula.terms.map { term =>
+      schema(term) match {
+        case column if column.dataType == StringType =>
+          val encodedTerm = term + "_onehot_" + uid
+          val indexer = factorLevels(term)
+          val indexCol = indexer.getOrDefault(indexer.outputCol)
+          encoderStages :+= indexer
+          encoderStages :+= new OneHotEncoder()
+            .setInputCol(indexCol)
+            .setOutputCol(encodedTerm)
+          tempColumns :+= encodedTerm
+          tempColumns :+= indexCol
+          encodedTerm
+        case _ =>
+          term
+      }
+    }
+    encoderStages :+= new VectorAssembler(uid)
+      .setInputCols(encodedTerms.toArray)
+      .setOutputCol($(featuresCol))
+    encoderStages :+= new ColumnPruner(tempColumns.toSet)
+    new PipelineModel(uid, encoderStages.toArray)
+  }
+}
+
+/**
+ * Utility transformer for removing temporary columns from a DataFrame.
+ */
+private class ColumnPruner(columnsToPrune: Set[String]) extends Transformer {
+  override val uid = Identifiable.randomUID("columnPruner")
+  override def transform(dataset: DataFrame): DataFrame = {
+    var res: DataFrame = dataset
+    for (column <- columnsToPrune) {
+      res = res.drop(column)
```

Comment:
Calling `drop` one by one might increase the stack size. We can get output columns by `dataset.columns.toSet -- columnsToPrune` and then call `select` directly.